coast/CLIMATOLOGY.py

```diff
@@ -76,10 +84,122 @@ def make_climatology(
+            (i.e. [(3,5),(12, 2)] is Mar -> May, Dec -> Feb). Must be in chronological order.
+
+            Returns:
+                date_ranges (list): A list of tuples, each containing a start and end datetime.date object.
+        """
+        date_ranges = []
+        for y in set(years):
+            y = int(y)
+            for period in month_periods:
+                start = period[0]
+                end = period[1]
+                if start > end:
+                    begin_date = date(y, start, 1)
+                    end_day = calendar.monthrange(y + 1, end)[1]
+                    end_date = date(y + 1, end, end_day)
+                else:
+                    begin_date = date(y, start, 1)
+                    end_day = calendar.monthrange(y, end)[1]
+                    end_date = date(y, end, end_day)
+                date_ranges.append((begin_date, end_date))
+        return date_ranges
+
+    @staticmethod
+    def multiyear_averages(ds: xr.Dataset, month_periods: list, time_var: str = "time", time_dim: str = "t_dim"):
+        """Calculate multiyear means for all Data variables in a dataset between a given start and end month.
+
+        Args:
+            ds (xr.Dataset): xarray dataset containing data.
+            month_periods (list): A list containing tuples of start and end month integers.
+            (i.e. [(3,5),(12, 2)] is Mar -> May, Dec -> Feb). Must be in chronological order.
+            The Season class can be used for convenience (e.g. Season.WINTER, Season.All etc. )
+            time_var (str): String representing the time variable name within the dataset.
+            time_dim (str): String representing the time dimension name within the dataset.
+        returns:
+            ds_mean (xr.Dataset): A new dataset containing mean averages for each data variable across all years.
+            Indexed by the key 'year'.
+        """
+
+        time_dim_da = ds[f"{time_dim}"]
+        time_var_da = ds[f"{time_var}"]
+
+        # If time dimension isn't np.datetime64 but time variable is, then swap time variable to be the dimension.
+        # A datetime type is required for slicing data using xarray's sel() method.
+        if not np.issubdtype(time_dim_da.dtype, np.datetime64) and np.issubdtype(time_var_da.dtype, np.datetime64):
+            warn("Time dimension is not np.datatime64 but time variable is. Swapping time dimension for data variable.")
+            new_ds = ds.swap_dims({f"{time_dim}": f"{time_var}"})
+            time_dim = time_var
+        elif np.issubdtype(time_dim_da.dtype, np.datetime64):
+            # Continue without warning.
+            new_ds = ds
+        else:
+            # Slicing will most likely fail for non np.datetime64 datatypes.
+            warn("Neither time dimension or time variable data are np.datetime64. Time slicing may fail.")
+            new_ds = ds
+
+        # Get years of data.
+        data_years = list(new_ds[f"{time_dim}.year"].to_numpy())
```

Comment:
Casting a large numpy array to list tends to be a very slow operation. Is this something we absolutely have to do?

Reply:
Now just casting the new_ds[f"{time_dim}.year"].data to a list directly.